client/network/src/service/traits.rs

```diff
@@ -632,3 +637,141 @@ where
+///
+/// After the validation result has been received by `Notifications`, it prepares the
+/// substream for communication by initializing the necessary sinks and emits
+/// `NotificationStreamOpened` which informs the protocol that the remote peer is ready to receive
+/// notifications.
+///
+/// Two different flavors of sending options are provided:
+///  * synchronous sending ([`ProtocolNotificationService::send_sync_notification()`])
+///  * asynchronous sending ([`ProtocolNotificationService::send_async_notification()`])
+///
+/// The former is used by protocols that don't have the implementation ready for exercising
+/// backpressure and the latter for the protocols that can do it.
+///
+/// Both local and remote peer can close the substream at any time. Local peer can do so by calling
+/// [`ProtocolNotificationService::close_substream()`] which instrucs `Notifications` to close the
+/// substream. Remote closing the substream is indicated to the local peer by receiving
+/// `NotificationEvent::SubstreamClosed` event
+///
+/// In case the protocol must update its handshake while it's operating (such as updating the best
+/// block information), it can do so by calling [`ProtocolNotificationService::set_handshake()`]
+/// which instructs `Notifications` to update the handshake it stored during protocol
+/// initialization.
+///
+/// All peer events are multiplexed on the same incoming event stream from `Notifications` and thus
+/// each event carries a `PeerId` so the protocol knows whose information to update when receiving
+/// an event.
+#[async_trait::async_trait]
+pub trait NotificationService: Debug + Send {
+	/// Instruct `Notifications` to open a new substream for `peer`.
+	///
+	/// `dial_if_disconnected` informs `Notifications` whether to dial
+	// the peer if there is currently no active connection to it.
+	async fn open_substream(&mut self, peer: PeerId) -> Result<(), ()>;
+
+	/// Instruct `Notifications` to close substream for `peer`.
+	async fn close_substream(&mut self, peer: PeerId) -> Result<(), ()>;
+
+	/// Send synchronous `notification` to `peer`.
+	fn send_sync_notification(
+		&self,
+		peer: &PeerId,
+		notification: Vec<u8>,
+	) -> Result<(), error::Error>;
+
+	/// Send asynchronous `notification` to `peer`, allowing sender to exercise backpressure.
+	async fn send_async_notification(
+		&self,
+		peer: &PeerId,
+		notification: Vec<u8>,
+	) -> Result<(), error::Error>;
+
+	/// Set handshake for the notification protocol replacing the old handshake.
+	async fn set_hanshake(&mut self, handshake: Vec<u8>) -> Result<(), ()>;
+
+	/// Get next event from the `Notifications` event stream.
+	async fn next_event(&mut self) -> Option<NotificationEvent>;
+
+	/// Make a copy of the object so it can be shared between protocol components
+	/// who wish to have access to the same underlying notification protocol.
+	fn clone(&mut self) -> Result<Box<dyn NotificationService>, ()>;
```

Comment:
Cloning `NotificationService` turns it into basically a broadcast receiver. I don't like this design at all and it wasn't even in the original design document but there are a few GRANDPA tests run voter and observer simultaneously and require access to the same notification event stream. Making the handle clonable was the cleanest way of working around that issue but I hope that this can be removed in the future if network refactoring of GRANDPA allows it.

Reply:
Yeah, thanks for the explanation. I left the comment before reading the trait implementation and forget about it.